### PR TITLE
feat: Setting and getting default accounts

### PR DESCRIPTION
**Pull request by aterga**
# Motivation

This is the next step towards supporting the new concept of default accounts. The BE now allows authorized clients to set and get default accounts for each (anchor, origin).

# Changes

* Implemented `set_default_account` and `get_default_account`.
* `update_account_for_origin` initializes the default account in the `anchor_application_config` index while creating that account.

# Tests

## Coverage Summary
* Happy paths: Default retrieval, customization, reset, multi-origin
* Error cases: Non-existent anchor/origin/account
* Edge cases: Account state changes, operation sequencing 

### get_default_account_for_origin Tests

1. `should_get_default_account_for_origin` - Comprehensive integration test covering:
    * GET without SET (returns synthetic default)
    * SET to existing accounts + GET (customizes default)
    * SET to non-existent accounts + GET (SET fails, GET unchanged)
    * SET to None + GET (resets to synthetic default)
2. `can_get_default_before_update_account_for_origin` - Returns the synthetic account for uninitialized origins.
3. `should_get_updated_default_account_after_modification` - Returns updated default account with assigned account number after modification.
4. `should_fail_get_default_account_for_nonexistent_anchor` - Returns the synthetic account for non-existent anchors.
5. `should_get_default_account_for_different_origins` - Each origin has independent default accounts.

### set_default_account_for_origin Tests

Test scenarios:

✅ SET to existing accounts (succeeds)
✅ SET to non-existent accounts (fails with NoSuchAccount)
✅ SET to None (resets to synthetic default)
✅ State persistence across operations

### CODE REVIEW

File: src/internet_identity/src/account_management.rs [hunk 1]
```diff
@@ -724,3 +810,284 @@ fn should_increment_discrepancy_counter() {
...
+            )),
+            Ok(AccountInfo {
+                account_number: Some(2),
+                origin: origin.clone(),
+                last_used: None,
+                name: Some("Bob".to_string()),
+            }),
+        ),
+        (
+            "SET to None (reset to default), then GET (can set default to None again)",
```
Comment (danblackadder) [inline src/internet_identity/src/account_management.rs:949]: Is it possible to make this situation throw an error instead if None is passed as the account_number?

It doesn't make sense that I should be able to call `set` with `none` does it?

I know we are supporting `None` by default, as the default primary, but going forward should we continue to support this if we are trying to improve the situation is all I am thinking? From a UI perspective, there is never a case you would technically be able to call this anyway, other than if we either added a button to say "Remove all defaults" or the canister is directly called via the user outside of the UI

Comment (aterga) [reply src/internet_identity/src/account_management.rs:949]: Good question.

I think currently setting to None is the only way to set the primary account as the default. 

If we wanted to migrate primary accounts to, e.g., `Some(0)` — which we might be able to do, we would need to implement some orthogonal changes on the BE

This approach assumes that we don't yet want to do that migration but already want default accounts to be customizable.

File: src/internet_identity/internet_identity.did [hunk 2]
```diff
@@ -653,7 +653,7 @@ type SetDefaultAccountError = variant {
     };
     NoSuchAccount : record {
         anchor_number : UserNumber;
-        account_number : opt AccountNumber;
+        origin : FrontendHostname;
```
Comment (lmuntaner) [inline src/internet_identity/internet_identity.did:656]: This means you need to generate the frontend types again, right?

File: src/internet_identity/src/account_management.rs [hunk 3]
```diff
@@ -39,6 +44,87 @@ pub fn get_accounts_for_origin(
...
+}
+
+pub fn get_default_account_for_origin(
+    anchor_number: AnchorNumber,
+    origin: FrontendHostname,
+) -> Result<AccountInfo, GetDefaultAccountError> {
+    let Some(application_number) =
+        storage_borrow(|storage| storage.lookup_application_number_with_origin(&origin))
+    else {
+        return Err(GetDefaultAccountError::NoSuchOrigin { anchor_number });
```
Comment (lmuntaner) [inline src/internet_identity/src/account_management.rs]: We should not fail in this case. Applications are created only when accounts are created, but we might request the default account before creating them.

If no application exists, it means that no account was ever created, so it should send the "Primary Account". Btw, do we have a new name for this? We used to call them "Default account".

File: src/internet_identity/src/account_management.rs [hunk 4]
```diff
@@ -39,6 +44,87 @@ pub fn get_accounts_for_origin(
     storage_borrow(|storage| storage.list_accounts(anchor_number, origin))
 }
 
+fn try_read_account_info(
```
Comment (lmuntaner) [inline src/internet_identity/src/account_management.rs:52]: It would be good to have some comments to understand the intended functionality of this function.

File: src/internet_identity/src/account_management.rs [hunk 5]
```diff
@@ -39,6 +44,87 @@ pub fn get_accounts_for_origin(
...
+            account_str, anchor_number, origin
+        );
+
+        return Err(message);
+    };
+
+    Ok(account.to_info())
+}
+
+pub fn get_default_account_for_origin(
```
Comment (lmuntaner) [inline src/internet_identity/src/account_management.rs:80]: Ditto adding some comments to understand it.

File: src/internet_identity/src/account_management.rs [hunk 6]
```diff
@@ -39,6 +44,87 @@ pub fn get_accounts_for_origin(
...
+    Ok(account)
+}
+
+pub fn set_default_account_for_origin(
+    anchor_number: AnchorNumber,
+    origin: FrontendHostname,
+    account_number: Option<AccountNumber>,
+) -> Result<AccountInfo, SetDefaultAccountError> {
+    let Some(application_number) =
+        storage_borrow(|storage| storage.lookup_application_number_with_origin(&origin))
```
Comment (lmuntaner) [inline src/internet_identity/src/account_management.rs]: We should use lookup_or_insert here, shouldn't we? Applications are created when accounts are created, and here we might be creating an account by setting it as default.

File: src/internet_identity/src/account_management.rs [hunk 7]
```diff
@@ -724,3 +810,284 @@ fn should_increment_discrepancy_counter() {
...
+
+        assert_eq!(
+            result, expected_get_result,
+            "{label}: unexpected GET result"
+        );
+    }
+}
+
+#[test]
+fn cannot_get_default_before_update_account_for_origin() {
```
Comment (lmuntaner) [inline src/internet_identity/src/account_management.rs]: Why this functionality? Accounts are created when non-default data needs to be stored, therefore, getting accounts before creating them just returns the (old) default account ("Primary account").

See read_account logic, that might be helpful as well.

File: src/internet_identity/src/storage.rs [hunk 8]
```diff
@@ -1232,6 +1278,17 @@ impl<M: Memory + Clone> Storage<M> {
         // Get or create an application number from the account's origin.
         let application_number =
             self.lookup_or_insert_application_number_with_origin(&params.origin);
+
+        // Update default account in the (anchor, origin) config.
```
Comment (lmuntaner) [inline src/internet_identity/src/storage.rs:1282]: This "Default account" is the "Primary Account", it doesn't have the same meaning as your "Default Account". We might want to rename this function...

Therefore, I don't think you should update the config here.

Default accounts are created for example updating the name of it. It might be that the user wants to rename from "Primary account" to another name. Those account don't necessarily exist in the storage, so we need to first create them before updating them.

See "update_account" for an example.

File: src/internet_identity/src/storage/account.rs [hunk 9]
```diff
@@ -65,7 +65,7 @@ pub struct Account {
     pub origin: FrontendHostname,
     pub last_used: Option<Timestamp>,
     pub name: Option<String>,
-    seed_from_anchor: Option<AnchorNumber>,
+    pub(crate) seed_from_anchor: Option<AnchorNumber>,
```
Comment (lmuntaner) [inline src/internet_identity/src/storage/account.rs]: Why is this needed?

File: src/internet_identity/internet_identity.did (hunk 2 above)
Comment (aterga) [reply src/internet_identity/internet_identity.did:656]: Yes, good catch

File: src/internet_identity/src/account_management.rs (hunk 6 above)
Comment (aterga) [reply src/internet_identity/src/account_management.rs]: Currently, `get_default_account_for_origin` is an update function, so we could do that,

but it should probably be changed to query. In which case we won't be able to modify the state.

File: src/internet_identity/src/account_management.rs (hunk 7 above)
Comment (aterga) [reply src/internet_identity/src/account_management.rs]: Yes, this is going to change. Thanks for confirming the intended behavior in this case.

File: src/internet_identity/src/storage.rs (hunk 8 above)
Comment (aterga) [reply src/internet_identity/src/storage.rs:1282]: Yes, we agreed with @sea-snake to do the renaming in a separate PR. 

For now, the terminology will be confusing for new readers, but at least we're all on the same page.

File: src/internet_identity/src/account_management.rs (hunk 6 above)
Comment (lmuntaner) [reply src/internet_identity/src/account_management.rs]: This is `set_default_account_for_origin`, it will always be an `update`, no?

In `get_default_account_for_origin` we shuold assume there are no accounts and return the "Primary Account" (shuold we start calling it that?)

Comment (aterga) [reply src/internet_identity/src/account_management.rs]: > "Primary Account" (shuold we start calling it that?)

I prefer "synthetic" as it makes it clear it's synthesized and not stored. 

Starting to use the new convention in this PR already, let me know if you disagree.

File: src/internet_identity/src/account_management.rs (hunk 7 above)
Comment (aterga) [reply src/internet_identity/src/account_management.rs]: Changed this to `can_get_default_before_update_account_for_origin`

File: src/internet_identity/src/account_management.rs (hunk 6 above)
Comment (lmuntaner) [reply src/internet_identity/src/account_management.rs]: Sounds good!

File: src/internet_identity/src/storage/account.rs (hunk 9 above)
Comment (aterga) [reply src/internet_identity/src/storage/account.rs]: Good catch, it isn't needed. Undoing the change.

File: src/internet_identity/internet_identity.did [hunk 10]
```diff
@@ -1011,7 +1011,7 @@ service : (opt InternetIdentityInit) -> {
     get_default_account : (
         anchor_number : UserNumber,
         origin : FrontendHostname,
-    ) -> (variant { Ok : AccountInfo; Err: GetDefaultAccountError });
+    ) -> (variant { Ok : AccountInfo; Err: GetDefaultAccountError }) query;
```
Comment (lmuntaner) [inline src/internet_identity/internet_identity.did:1014]: Did you add this before or after the declarations? I'm not sure, but I think that the code generated changes also with query calls.

File: src/internet_identity/src/account_management.rs [hunk 11]
```diff
@@ -39,6 +44,108 @@ pub fn get_accounts_for_origin(
...
+            account_number, anchor_number, origin
+        );
+
+        return Err(message);
+    };
+
+    Ok(account.to_info())
+}
+
+/// Best effort to determin the default account for the given (anchor, origin).
```
Comment (lmuntaner) [inline src/internet_identity/src/account_management.rs:77]: Small typo "determin".

Comment (lmuntaner) [reply src/internet_identity/src/account_management.rs:77]: I would add in the comment the scenarios:

* If application not found: synthetic account.
* If no default account stored: synthetic account.
* Else, read from storage, raise error if not found.

File: src/internet_identity/src/storage/storable/anchor_application_config.rs [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+use ic_stable_structures::storable::Bound;
+use ic_stable_structures::Storable;
+use minicbor::{Decode, Encode};
+use std::borrow::Cow;
+
+#[derive(Encode, Decode, Default, Clone, Ord, Eq, PartialEq, PartialOrd)]
+#[cbor(map)]
+pub struct AnchorApplicationConfig {
+    #[n(0)]
+    pub default_account_number: Option<StorableAccountNumber>, // None is the unreserved default account
```
Comment (lmuntaner) [inline src/internet_identity/src/storage/storable/anchor_application_config.rs:11]: Change "None is the unreserved default account" for "the unreserved synthetic account"?

File: src/internet_identity/src/account_management.rs [hunk 13]
```diff
@@ -39,6 +44,108 @@ pub fn get_accounts_for_origin(
...
+        application_number,
+        default_account_number,
+    )
+    .map_err(GetDefaultAccountError::InternalCanisterError)?;
+
+    Ok(account)
+}
+
+/// Sets the default account for the given (anchor, origin) to the specified `account_number`.
+///
```
Comment (lmuntaner) [inline src/internet_identity/src/account_management.rs:112]: I would add a comment that if the account_number is None, then the synthetic account is returned and also `None` is stored.

File: src/internet_identity/src/main.rs [hunk 14]
```diff
@@ -388,31 +388,33 @@ fn update_account(
...
 
 #[update]
 fn set_default_account(
     anchor_number: AnchorNumber,
-    _origin: FrontendHostname,
-    _account_number: Option<AccountNumber>,
+    origin: FrontendHostname,
+    account_number: Option<AccountNumber>,
 ) -> Result<AccountInfo, SetDefaultAccountError> {
     check_authorization(anchor_number)
```
Comment (lmuntaner) [inline src/internet_identity/src/main.rs:411]: I remembered now that we have a helper `anchor_operation_with_authz_check`. Worth doing in another PR though.

The idea is that all update operations are stored then in the archive canister.

Which also means that this operation or setting an account as default must be stored there as well.

File: src/internet_identity/src/storage.rs (hunk 8 above)
Comment (lmuntaner) [reply src/internet_identity/src/storage.rs:1282]: I would add in the comment of `create_default_account` that the account will be set as default as well.

Did you think of all the use cases for this function that it makes sense to set it as default account as well? Isn't there an edge case where another default account already exists and we don't want to substitute it?

Maybe setting the default should be done where this is called instead?

File: src/internet_identity/src/storage/account.rs [hunk 15]
```diff
@@ -69,6 +69,18 @@ pub struct Account {
 }
 
 impl Account {
+    /// A default account that is not meant to be stored.
```
Comment (lmuntaner) [inline src/internet_identity/src/storage/account.rs:72]: I would not call it "default"... hehe.

File: src/internet_identity/src/storage/account.rs [hunk 16]
```diff
@@ -69,6 +69,18 @@ pub struct Account {
...
+    pub fn synthetic(anchor_number: AnchorNumber, origin: FrontendHostname) -> Self {
```
Comment (lmuntaner) [inline src/internet_identity/src/storage/account.rs:73]: Great idea!!!

**Review by lmuntaner — APPROVED** [review]
Some more comments, but already approving!

File: src/internet_identity/internet_identity.did (hunk 10 above)
Comment (aterga) [reply src/internet_identity/internet_identity.did:1014]: Good catch, [done](https://github.com/dfinity/internet-identity/pull/3382)

File: src/internet_identity/src/main.rs (hunk 14 above)
Comment (aterga) [reply src/internet_identity/src/main.rs:411]: Why isn't this used in `update_account`?

Comment (aterga) [reply src/internet_identity/src/main.rs:411]: Oh I see, it's just done directly via post_operation_bookkeeping
